include/pybind11/pytypes.h

```diff
@@ -375,22 +387,82 @@ PYBIND11_NAMESPACE_END(detail)
-    /// error variables (but the `.what()` string is still available).
+    const char *what() const noexcept override {
+        if (m_lazy_what.empty()) {
+            try {
+                m_lazy_what = detail::error_string(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+                // Negate the if condition to test the catch(...) block below.
+                if (m_lazy_what.empty()) {
+                    throw std::runtime_error(
+                        "FATAL failure building pybind11::detail::error_already_set what()");
+                }
+            } catch (...) {
+                // Terminating the process, to not mask the original error by errors in the error
+                // handling. Reporting the original error on stderr & stdout. Intentionally using
+                // the Python C API directly, to maximize reliability.
+                std::string msg
+                    = "FATAL failure building pybind11::detail::error_already_set what(): ";
+                if (m_type.ptr() == nullptr) {
+                    msg += "PYTHON_EXCEPTION_TYPE_IS_NULLPTR";
+                } else {
+                    const char *class_name = detail::obj_class_name(m_type.ptr());
+                    if (class_name == nullptr) {
+                        msg += "PYTHON_EXCEPTION_CLASS_NAME_IS_NULLPTR";
+                    } else {
+                        msg += class_name;
+                    }
+                }
+                msg += ": ";
+                PyObject *val_str = PyObject_Str(m_value.ptr());
+                if (val_str == nullptr) {
+                    msg += "PYTHON_EXCEPTION_VALUE_IS_NULLPTR";
+                } else {
+                    Py_ssize_t utf8_str_size = 0;
+                    const char *utf8_str = PyUnicode_AsUTF8AndSize(val_str, &utf8_str_size);
+                    if (utf8_str == nullptr) {
+                        msg += "PYTHON_EXCEPTION_VALUE_AS_UTF8_FAILURE";
+                    } else {
+                        msg += '"' + std::string(utf8_str, static_cast<std::size_t>(utf8_str_size))
+                               + '"';
+                    }
+                }
+                // Intentionally using C calls to maximize reliability
+                // (and to avoid #include <iostream>).
+                fprintf(stderr, "%s [STDERR]\n", msg.c_str());
+                fflush(stderr);
+                fprintf(stdout, "%s [STDOUT]\n", msg.c_str());
+                fflush(stdout);
+                std::terminate();
+            }
+        }
+        return m_lazy_what.c_str();
+    }
+
+    /// Restores the currently-held Python error (which will clear the Python error indicator first
+    //  if already set).
     void restore() {
-        PyErr_Restore(m_type.release().ptr(), m_value.release().ptr(), m_trace.release().ptr());
+        // As long as this type is copyable, there is no point in releasing m_type, m_value,
```

Comment:
The original idea was to have what() cached so that it still worked after err.restore was called. That's why we called what() in restore.